src/session/session_dhandle.c

```diff
@@ -405,12 +405,26 @@ __session_find_shared_dhandle(WT_SESSION_IMPL *session, const char *uri, const c
 {
     WT_DECL_RET;
 
-    WT_WITH_HANDLE_LIST_READ_LOCK(session,
-      if ((ret = __wt_conn_dhandle_find(session, uri, checkpoint)) == 0)
-        WT_DHANDLE_ACQUIRE(session->dhandle));
+    /*
+     * If we get called as part of the checkpoint, and the handle is the checkpoint to be created,
+     * it is guaranteed that the handle won't exist in the shared dhandle list. Skip searching
+     * through the list, and directly insert into it.
+     */
+    if (WT_SESSION_IS_CHECKPOINT(session) && checkpoint != NULL &&
+      WT_STRING_MATCH(checkpoint, WT_CHECKPOINT, strlen(WT_CHECKPOINT))) {
+        /*
+         * Let's check that our assumption that this handle is not already there in the shared list
+         * is actually correct.
+         */
+        WT_ASSERT(session, __wt_conn_dhandle_find(session, uri, checkpoint) != 0);
```

Comment:
This call still needs the `WT_WITH_HANDLE_LIST_READ_LOCK` or it could follow changing pointers and segfault. Also I think the assert should check explicitly for `WT_NOTFOUND`. Any other error is a real error. So it should be:
`WT_WITH...(ret = wt_conn_dhandle_find(...))`
`WT_ASSERT(session, ret != WT_NOTFOUND);`

That means it isn't skipped overall though, so I think this code is going to need to be under `HAVE_DIAGNOSTIC` instead. I think the code overall would read better if the conditional were flipped and the `if` portion is NOT a checkpoint. Then the `else` would be the diagnostic code.

Reply:
Thanks @sueloverso for the comments. Yeah, even I found some issue with the current change, so changed the status of the PR to draft. 
I will address the review comments next Monday as I am working on a Skunkwork project this week.